File: src/gpio/gpio.service.ts
// src/gpio/gpio.service.ts

import {BadRequestException, HttpStatus, Injectable, Logger, OnModuleDestroy, UnauthorizedException} from '@nestjs/common';
import { Gpio } from 'onoff';
import { ApiService } from '../api/api.service';
import { UserService } from '../user/user.service';
import { BULB_PINS, SWITCH_PINS } from '../constants/pin.constants';
import * as fs from 'fs';

@Injectable()
export class GPIOService implements OnModuleDestroy {
  private readonly logger = new Logger(GPIOService.name);

  // GPIO instances
  private switches: Gpio[] = [];
  private bulbs: Gpio[] = [];

  // State tracking
  private switchStates: boolean[] = new Array(5).fill(false);
  private isMonitoring = false;
  private monitoringIntervals: ReturnType<typeof setTimeout>[] = [];
  private gpioAvailable = false;

  constructor(
    private readonly apiService: ApiService,
    private readonly userService: UserService,
  ) {}

  async initialize() {
    try {
      this.logger.log('🔍 Checking GPIO availability...');

      // CRITICAL: Check GPIO availability FIRST
      this.gpioAvailable = this.checkGpioAvailability();

      if (!this.gpioAvailable) {
        this.logger.warn('⚠️ GPIO hardware not available - activating simulation mode');
        this.logger.log('💡 Test switches using: POST /test/switch/{0-4}');
        this.logger.log('🔧 All GPIO operations will be simulated');
        return; // Exit early - no hardware initialization
      }

      this.logger.log('🔌 GPIO hardware detected - initializing physical pins...');

      // Only initialize hardware if GPIO is available
      await this.initializeHardware();

      this.logger.log('✅ GPIO hardware initialization completed');
    } catch (error) {
      this.logger.error('❌ GPIO initialization failed:', error);
      this.gpioAvailable = false; // Fallback to simulation
      this.logger.warn('🔄 Falling back to simulation mode due to hardware error');
      this.cleanup();
    }
  }

  private async initializeHardware() {
    // Initialize switch pins (input with pull-up)
    for (let i = 0; i < SWITCH_PINS.length; i++) {
      const pin = SWITCH_PINS[i];
      try {
        const gpio = new Gpio(pin, 'in', 'rising', { activeLow: true });
        this.switches.push(gpio);
        this.logger.log(`✅ Switch ${i + 1} initialized on GPIO ${pin}`);
      } catch (error) {
        this.logger.error(`❌ Failed to initialize switch ${i + 1} on GPIO ${pin}:`, error);
        throw error;
      }
    }

    // Initialize bulb pins (output)
    for (let i = 0; i < BULB_PINS.length; i++) {
      const pin = BULB_PINS[i];
      try {
        const gpio = new Gpio(pin, 'out');
        await gpio.write(0); // Initially off
        this.bulbs.push(gpio);
        this.logger.log(`✅ Bulb ${i + 1} initialized on GPIO ${pin}`);
      } catch (error) {
        this.logger.error(`❌ Failed to initialize bulb ${i + 1} on GPIO ${pin}:`, error);
        throw error;
      }
    }
  }

  private checkGpioAvailability(): boolean {
    try {
      // Check basic GPIO requirements
      const hasGpioExport = fs.existsSync('/sys/class/gpio/export');
      const hasGpioMem = fs.existsSync('/dev/gpiomem');
      const hasDeviceTree = fs.existsSync('/proc/device-tree/model');

      this.logger.log(`🔍 GPIO Check - Export: ${hasGpioExport}, Memory: ${hasGpioMem}, DeviceTree: ${hasDeviceTree}`);

      if (!hasGpioExport || !hasGpioMem || !hasDeviceTree) {
        return false;
      }

      // Check if it's actually a Raspberry Pi
      try {
        const model = fs.readFileSync('/proc/device-tree/model', 'utf8');
        const isRaspberryPi = model.includes('Raspberry Pi');
        this.logger.log(`📱 Device: ${model.trim()}`);
        this.logger.log(`🍓 Is Raspberry Pi: ${isRaspberryPi}`);

        if (!isRaspberryPi) {
          return false;
        }
      } catch (error) {
        this.logger.warn(`⚠️ Could not read device model: ${error.message}`);
        return false;
      }

      // Test write access to GPIO export
      try {
        fs.accessSync('/sys/class/gpio/export', fs.constants.W_OK);
        this.logger.log('✅ GPIO export is writable');
        return true;
      } catch (writeError) {
        this.logger.warn(`⚠️ GPIO export not writable: ${writeError.message}`);
        return false;
      }

    } catch (error) {
      this.logger.warn(`⚠️ GPIO availability check failed: ${error.message}`);
      return false;
    }
  }

  startMonitoring() {
    if (this.isMonitoring) {
      this.logger.warn('⚠️ GPIO monitoring already active');
      return;
    }

    this.logger.log('🚀 Starting GPIO monitoring...');
    this.isMonitoring = true;

    if (!this.gpioAvailable) {
      this.logger.log('📝 SIMULATION MODE: GPIO monitoring active');
      this.logger.log('🧪 Test endpoints: POST /test/switch/{0-4}');
      return;
    }

    // Set up interrupt-based monitoring for each switch
    this.switches.forEach((gpio, index) => {
      gpio.watch((err, value) => {
        if (err) {
          this.logger.error(`❌ Error monitoring switch ${index + 1}:`, err);
          return;
        }

        if (value === 1) {
          this.handleSwitchPress(index);
        }
      });
    });

    this.logger.log('✅ HARDWARE MODE: Physical GPIO monitoring active');
  }

  async handleSwitchPress(switchIndex: number) {
    if (switchIndex <= 0 || switchIndex >= SWITCH_PINS.length) throw new BadRequestException(`Invalid switch position: ${switchIndex}`);

    this.logger.log(`🔘 Switch ${switchIndex + 1} activated`);

    try {
      const promises = [
        this.turnOnBulb(SWITCH_PINS[switchIndex - 1]),
        this.sendApiRequest(SWITCH_PINS[switchIndex - 1])
      ];

      const [turnOnBulb, sendApiRequest ] = await Promise.all(promises);
      return { turnOnBulb, sendApiRequest };
    } catch (error) {
      if (error.status === HttpStatus.UNAUTHORIZED) throw new UnauthorizedException(error.response.data);
    }
  }

  private async turnOnBulb(bulbIndex: number) {
    try {
      if (!this.gpioAvailable) {
        // Simulation mode
        this.logger.log(`🔧 SIMULATION: Bulb ${bulbIndex + 1} → ON`);
        await new Promise(resolve => setTimeout(resolve, 100));
        this.logger.log(`💡 SIMULATION: Bulb ${bulbIndex + 1} illuminated (2s)`);
        await new Promise(resolve => setTimeout(resolve, 2000));
        this.logger.log(`🔧 SIMULATION: Bulb ${bulbIndex + 1} → OFF`);
        return;
      }

      // Hardware mode
      const bulb = this.bulbs[bulbIndex];
      if (!bulb) {
        this.logger.error(`❌ Bulb ${bulbIndex + 1} not initialized`);
        return;
      }

      await bulb.write(1);
      this.logger.log(`💡 HARDWARE: Bulb ${bulbIndex + 1} → ON (GPIO ${BULB_PINS[bulbIndex]})`);

      await new Promise(resolve => setTimeout(resolve, 2000));

      await bulb.write(0);
      this.logger.log(`💡 HARDWARE: Bulb ${bulbIndex + 1} → OFF (GPIO ${BULB_PINS[bulbIndex]})`);

    } catch (error) {
      this.logger.error(`❌ Error controlling bulb ${bulbIndex + 1}:`, error);
    }
  }

  private async sendApiRequest(switchIndex: number) {
    // try {
      const user = await this.userService.getUser(switchIndex);

      if (!user) {
        this.logger.error(`❌ No user found for switch ${switchIndex + 1}`);
        return;
      }

      const payload = {
        status: 'calling',
        branchId: process.env.DEVICE_ID || '',
        isMultiService: false,
        location: user.location,
      };

      this.logger.log(`📡 API: Sending request for switch ${switchIndex + 1} (user: ${user.userId})`);
      const {data} = await this.apiService.sendSwitchEvent(payload, user.accessToken);

      this.logger.log(`✅ API: Request completed for switch ${switchIndex + 1}`);

      return data?.data;
    // } catch (error) {
    //   this.logger.error(`❌ API: Failed for switch ${switchIndex + 1}:`);
    // }
  }

  getStatus() {
    return {
      gpioAvailable: this.gpioAvailable,
      mode: this.gpioAvailable ? 'hardware' : 'simulation',
      isMonitoring: this.isMonitoring,
      switchStates: this.switchStates,
      switchPins: SWITCH_PINS,
      bulbPins: BULB_PINS,
      switchCount: this.switches.length,
      bulbCount: this.bulbs.length,
      timestamp: new Date().toISOString(),
      systemInfo: {
        hasGpioExport: fs.existsSync('/sys/class/gpio/export'),
        hasGpioMem: fs.existsSync('/dev/gpiomem'),
        hasDeviceTree: fs.existsSync('/proc/device-tree/model'),
      }
    };
  }

  onModuleDestroy() {
    this.cleanup();
  }

  private cleanup() {
    this.logger.log('🧹 Cleaning up GPIO resources...');
    this.isMonitoring = false;

    this.monitoringIntervals.forEach(interval => clearInterval(interval));
    this.monitoringIntervals = [];

    if (!this.gpioAvailable) {
      this.logger.log('✅ Cleanup completed (simulation mode)');
      return;
    }

    // Hardware cleanup
    this.switches.forEach((gpio, index) => {
      try {
        gpio.unexport();
        this.logger.log(`✅ Switch ${index + 1} cleaned up`);
      } catch (error) {
        this.logger.error(`⚠️ Error cleaning switch ${index + 1}:`, error);
      }
    });

    this.bulbs.forEach((gpio, index) => {
      try {
        gpio.writeSync(0);
        gpio.unexport();
        this.logger.log(`✅ Bulb ${index + 1} cleaned up`);
      } catch (error) {
        this.logger.error(`⚠️ Error cleaning bulb ${index + 1}:`, error);
      }
    });

    this.switches = [];
    this.bulbs = [];
    this.logger.log('✅ Hardware cleanup completed');
  }
}

File: app/services/gpio_service.py
# app/services/gpio_service.py

import asyncio
import os
from datetime import datetime
from typing import List, Dict, Any
from loguru import logger

from app.constants.pin_constants import SWITCH_PINS, BULB_PINS
from app.models.models import SwitchEventPayload, UserLocation
from app.services.api_service import ApiService
from app.services.user_service import UserService
from app.core.config import settings

try:
    import RPi.GPIO as GPIO
    GPIO_AVAILABLE = True
    logger.info("📦 RPi.GPIO library loaded successfully")
except ImportError:
    GPIO_AVAILABLE = False
    logger.warning("⚠️ RPi.GPIO not available - running in simulation mode")
    logger.info("💡 This is normal on non-Raspberry Pi systems")

    # Create a mock GPIO class for development
    class MockGPIO:
        BCM = "BCM"
        IN = "IN"
        OUT = "OUT"
        HIGH = 1
        LOW = 0
        PUD_UP = "PUD_UP"

        @staticmethod
        def setmode(mode): pass
        @staticmethod
        def setwarnings(warnings): pass
        @staticmethod
        def setup(pin, mode, pull_up_down=None): pass
        @staticmethod
        def input(pin): return 0
        @staticmethod
        def output(pin, value): pass
        @staticmethod
        def cleanup(): pass

    GPIO = MockGPIO()


class GPIOService:
    def __init__(self):
        self.api_service = ApiService()
        self.user_service = UserService()

        # GPIO state
        self.gpio_available = False
        self.is_monitoring = False
        self.switch_states = [False] * 5

        # Monitoring tasks
        self.monitoring_tasks: List[asyncio.Task] = []

    async def initialize(self) -> None:
        """Initialize GPIO service"""
        try:
            logger.info("🔍 Checking GPIO availability...")

            # Check GPIO availability first
            self.gpio_available = await self._check_gpio_availability()

            if not self.gpio_available:
                logger.warning("⚠️ GPIO hardware not available - activating simulation mode")
                logger.info("💡 Test switches using: GET /test/switch/{1-5}")
                logger.info("🔧 All GPIO operations will be simulated")
                return

            logger.info("🔌 GPIO hardware detected - initializing physical pins...")
            await self._initialize_hardware()
            logger.info("✅ GPIO hardware initialization completed")

        except Exception as error:
            logger.error(f"❌ GPIO initialization failed: {error}")
            self.gpio_available = False
            logger.warning("🔄 Falling back to simulation mode due to hardware error")
            await self.cleanup()

    async def _initialize_hardware(self) -> None:
        """Initialize GPIO hardware"""
        if not GPIO_AVAILABLE:
            return

        # Setup GPIO mode
        GPIO.setmode(GPIO.BCM)
        GPIO.setwarnings(False)

        # Initialize switch pins (input with pull-up)
        for i, pin in enumerate(SWITCH_PINS):
            try:
                GPIO.setup(pin, GPIO.IN, pull_up_down=GPIO.PUD_UP)
                logger.info(f"✅ Switch {i + 1} initialized on GPIO {pin}")
            except Exception as error:
                logger.error(f"❌ Failed to initialize switch {i + 1} on GPIO {pin}: {error}")
                raise

        # Initialize bulb pins (output)
        for i, pin in enumerate(BULB_PINS):
            try:
                GPIO.setup(pin, GPIO.OUT)
                GPIO.output(pin, GPIO.LOW)  # Initially off
                logger.info(f"✅ Bulb {i + 1} initialized on GPIO {pin}")
            except Exception as error:
                logger.error(f"❌ Failed to initialize bulb {i + 1} on GPIO {pin}: {error}")
                raise

    async def _check_gpio_availability(self) -> bool:
        """Check if GPIO is available"""
        try:
            if not GPIO_AVAILABLE:
                return False

            # Check if running in Docker
            is_docker = self._is_running_in_docker()
            if is_docker:
                logger.info("🐋 Running in Docker container")

            # Check basic GPIO requirements
            has_gpio_export = os.path.exists("/sys/class/gpio/export")
            has_gpio_mem = os.path.exists("/dev/gpiomem")
            has_device_tree = os.path.exists("/proc/device-tree/model")

            logger.info(f"🔍 GPIO Check - Export: {has_gpio_export}, Memory: {has_gpio_mem}, DeviceTree: {has_device_tree}")

            if not has_gpio_export or not has_gpio_mem:
                return False

            # Check if it's actually a Raspberry Pi
            if has_device_tree:
                try:
                    with open("/proc/device-tree/model", "r") as f:
                        model = f.read().strip()

                    is_raspberry_pi = "Raspberry Pi" in model
                    logger.info(f"📱 Device: {model}")
                    logger.info(f"🍓 Is Raspberry Pi: {is_raspberry_pi}")

                    if not is_raspberry_pi and not is_docker:
                        return False

                except Exception as error:
                    logger.warning(f"⚠️ Could not read device model: {error}")

            # Test GPIO access
            try:
                # Test if we can access GPIO (this will fail if not on Pi or in Docker without privileges)
                GPIO.setmode(GPIO.BCM)
                GPIO.setwarnings(False)
                logger.info("✅ GPIO access test successful")
                return True

            except Exception as error:
                logger.warning(f"⚠️ GPIO access test failed: {error}")
                return False

        except Exception as error:
            logger.warning(f"⚠️ GPIO availability check failed: {error}")
            return False

    def _is_running_in_docker(self) -> bool:
        """Check if running in Docker container"""
        try:
            # Check for .dockerenv file
            if os.path.exists("/.dockerenv"):
                return True

            # Check cgroup
            with open("/proc/1/cgroup", "r") as f:
                cgroup = f.read()

            return "docker" in cgroup or "containerd" in cgroup

        except Exception:
            return False

    async def start_monitoring(self) -> None:
        """Start GPIO monitoring"""
        if self.is_monitoring:
            logger.warning("⚠️ GPIO monitoring already active")
            return

        logger.info("🚀 Starting GPIO monitoring...")
        self.is_monitoring = True

        if not self.gpio_available:
            logger.info("📝 SIMULATION MODE: GPIO monitoring active")
            logger.info("🧪 Test endpoints: GET /test/switch/{1-5}")
            return

        # Start monitoring task for each switch
        for i in range(len(SWITCH_PINS)):
            task = asyncio.create_task(self._monitor_switch(i))
            self.monitoring_tasks.append(task)

        logger.info("✅ HARDWARE MODE: Physical GPIO monitoring active")

    async def _monitor_switch(self, switch_index: int) -> None:
        """Monitor a specific switch for state changes"""
        if not GPIO_AVAILABLE:
            return

        pin = SWITCH_PINS[switch_index]
        previous_state = GPIO.input(pin)

        while self.is_monitoring:
            try:
                current_state = GPIO.input(pin)

                # Detect rising edge (button press)
                if previous_state == GPIO.HIGH and current_state == GPIO.LOW:
                    await self.handle_switch_press(switch_index + 1)

                previous_state = current_state
                await asyncio.sleep(0.01)  # 10ms polling interval

            except Exception as error:
                logger.error(f"❌ Error monitoring switch {switch_index + 1}: {error}")
                break

    async def handle_switch_press(self, switch_index: int) -> Dict[str, Any]:
        """Handle switch press event"""
        if switch_index <= 0 or switch_index > len(SWITCH_PINS):
            raise ValueError(f"Invalid switch position: {switch_index}")

        logger.info(f"🔘 Switch {switch_index} activated")

        try:
            # Execute bulb control and API request concurrently
            bulb_task = asyncio.create_task(self._turn_on_bulb(switch_index - 1))
            api_task = asyncio.create_task(self._send_api_request(SWITCH_PINS[switch_index - 1]))

            bulb_result, api_result = await asyncio.gather(bulb_task, api_task)

            return {
                "turnOnBulb": bulb_result,
                "sendApiRequest": api_result
            }

        except Exception as error:
            logger.error(f"❌ Error handling switch press: {error}")
            raise

    async def _turn_on_bulb(self, bulb_index: int) -> Dict[str, Any]:
        bulb = bulb_index + 1

        """Turn on bulb for 2 seconds"""
        try:
            if not self.gpio_available:
                # Simulation mode
                logger.info(f"🔧 SIMULATION: Bulb {bulb} → ON")
                await asyncio.sleep(0.1)
                logger.info(f"💡 SIMULATION: Bulb {bulb} illuminated (2s)")
                await asyncio.sleep(settings.time_on_bulb)
                logger.info(f"🔧 SIMULATION: Bulb {bulb} → OFF")
                return {"simulated": True, "bulb": bulb}

            # Hardware mode
            if not GPIO_AVAILABLE:
                return {"error": "GPIO not available"}

            pin = BULB_PINS[bulb_index]

            GPIO.output(pin, GPIO.HIGH)
            logger.info(f"💡 HARDWARE: Bulb {bulb} → ON (GPIO {pin})")

            await asyncio.sleep(settings.time_on_bulb)

            GPIO.output(pin, GPIO.LOW)
            logger.info(f"💡 HARDWARE: Bulb {bulb} → OFF (GPIO {pin})")

            return {"hardware": True, "bulb": bulb, "gpio": pin}

        except Exception as error:
            logger.error(f"❌ Error controlling bulb {bulb}: {error}")
            return {"error": str(error)}

    async def _send_api_request(self, switch_index: int) -> Dict[str, Any]:
        """Send API request for switch press"""
        try:
            user = self.user_service.get_user(switch_index)

            if not user:
                logger.error(f"❌ No user found for switch pin {switch_index}")
                return {"error": "No user found"}

            payload = SwitchEventPayload(
                status="calling",
                branchId=settings.device_id,
                isMultiService=False,
                location=UserLocation(**user["location"])
            )

            logger.info(f"📡 API: Sending request for switch pin {switch_index} (user: {user['userId']})")
            response = await self.api_service.send_switch_event(payload, user["accessToken"])

            logger.info(f"✅ API: Request completed for switch pin {switch_index}")
            return response.get("data", response)

        except Exception as error:
            logger.error(f"❌ API: Failed for switch pin {switch_index}: {error}")
            raise

    def get_status(self) -> Dict[str, Any]:
        """Get GPIO service status"""
        return {
            "gpioAvailable": self.gpio_available,
            "mode": "hardware" if self.gpio_available else "simulation",
            "isMonitoring": self.is_monitoring,
            "isDocker": self._is_running_in_docker(),
            "switchStates": self.switch_states,
            "switchPins": SWITCH_PINS,
            "bulbPins": BULB_PINS,
            "switchCount": len(SWITCH_PINS),
            "bulbCount": len(BULB_PINS),
            "timestamp": datetime.now().isoformat(),
            "systemInfo": {
                "hasGpioExport": os.path.exists("/sys/class/gpio/export"),
                "hasGpioMem": os.path.exists("/dev/gpiomem"),
                "hasDeviceTree": os.path.exists("/proc/device-tree/model"),
            }
        }

    async def cleanup(self) -> None:
        """Cleanup GPIO resources"""
        logger.info("🧹 Cleaning up GPIO resources...")
        self.is_monitoring = False

        # Cancel monitoring tasks
        for task in self.monitoring_tasks:
            if not task.done():
                task.cancel()

        if self.monitoring_tasks:
            await asyncio.gather(*self.monitoring_tasks, return_exceptions=True)

        self.monitoring_tasks.clear()

        if not self.gpio_available or not GPIO_AVAILABLE:
            logger.info("✅ Cleanup completed (simulation mode)")
            return

        try:
            # Turn off all bulbs
            for i, pin in enumerate(BULB_PINS):
                try:
                    GPIO.output(pin, GPIO.LOW)
                    logger.info(f"✅ Bulb {i + 1} turned off")
                except Exception as error:
                    logger.error(f"⚠️ Error turning off bulb {i + 1}: {error}")

            # Cleanup GPIO
            GPIO.cleanup()
            logger.info("✅ Hardware cleanup completed")

        except Exception as error:
            logger.error(f"⚠️ Error during GPIO cleanup: {error}")
